_posts/2019-01-10-future-workloads.md

```diff
@@ -0,0 +1,500 @@
+
+```
+ms
+|1------------------------------------------Inf|
+```
+
+#### Implicitly or explicitly
+
+We looked at the `Future.apply` method definition before. It expects an implicit `ExecutionContext`, and this makes it tempting to always pass it implicitly in our applications.
+
+For example, here we import it at top level and all our futures will pick it up:
+
+```
+object FutureApp extends App {
+    import scala.concurrent.ExecutionContext.Implicits.global
+
+    ...
+
+    Future(blockingMethod(1))
+
+    ...
+
+    Future(cpuBoundMethod(1))
+}
+```
+
+This may have consequences if that is not expected, for example when we want to use another executor for blocking.
+
+##### Explicitly
+
+To fix the above, we can obviously pass the blocking context explicitly and default the rest.
+
+```
+object FutureApp extends App {
+    val defaultExecutor: ExecutionContext =  scala.concurrent.ExecutionContext.Implicits.global
+    val blockingExecutor: ExecutionContext = ...
+
+    ...
+
+    Future(blockingMethod(1))(blockingExecutor)
+
+    ...
+
+    Future(cpuBoundMethod(1))
+}
+```
+
+This is a micro management approach that can work when dealing with small code bases or teams.
+
+##### Explicitly implicitly
+
+We can improve on the above by typing our futures and thus marking them as blocking. This makes reutilization safer.
+
+```
+val defaultExecutor: ExecutionContext =  scala.concurrent.ExecutionContext.Implicits.global
+val blockingExecutor: ExecutionContext = ...
+
+case class BlockingContext(executionContext: ExecutionContext)
+
+def markedBlocking(value: Int)(implicit blockingContext: BlockingContext) = {
+    implicit val ec = blockingContext.executionContext
+    Future(blockingMethod(value))
+}
+
+markedBlocking(10)(BlockingContext(blockingExecutor)) // compiles
+markedBlocking(10)(defaultExecutor) // does not compile
+```
+
+The `markedBlocking` method would be public and used in place of our initial `blockingMethod`. This approach solves the boilerplate and risk of passing executors explicitly, documents our code and makes it safer for larger teams to understand where code requires a different execution strategy.
+
+#### Interlude
+
+So far, we should have gained some intuition on how `Future` and `ExecutionContext` work when interacting with three types of workloads.
+
+We have seen how to manage `blocking`, which is a special case where we [delegate control to the OS](http://pages.cs.wisc.edu/~remzi/OSTEP/cpu-mechanisms.pdf).
+
+We have also seen how an `asynchronous` method interacts with a future. Our asynchronous method is actually a special case where no computation is involved in our program, often referred to as asynchronous IO.
+
+`cpu bound` remains as the most frequent workload a common application will probably have. We have shown [here](#cpu-bound-pool), [here](#cpu-bound) and [here](#thread-pools) how changing the running time and thus level of concurrency can impact our choices.
```

Comment:
This may be controverse, but I would like to know why you claim cpu bound to be probably the most common. Even if it's not to be included in the post.

Reply:
I would say it is because this type of workload includes any type of computation, such as math, building and mapping classes, if elses, etc. The other two presented are the special cases of blocking calls (which have a kernel call somewhere inside) and async IO (delegate computation somewhere else)